pkg/types/azure/validation/machinepool.go

```diff
@@ -221,6 +235,91 @@ func validateOSImage(p *azure.MachinePool, fldPath *field.Path) field.ErrorList
+// configuration conflicts and ensure consistent security posture.
+
+func getDataDisksSecurityProfileType(p *azure.MachinePool) capz.SecurityEncryptionType {
+	for _, dataDisk := range p.DataDisks {
+		if dataDisk.ManagedDisk != nil && dataDisk.ManagedDisk.SecurityProfile != nil {
+			switch dataDisk.ManagedDisk.SecurityProfile.SecurityEncryptionType {
+			case capz.SecurityEncryptionTypeDiskWithVMGuestState:
+				return capz.SecurityEncryptionTypeDiskWithVMGuestState
+			case capz.SecurityEncryptionTypeVMGuestStateOnly:
+				return capz.SecurityEncryptionTypeVMGuestStateOnly
+
+			default:
+				continue
+			}
+		}
+	}
+	return ""
+}
+
+func validateDataDisk(p *azure.MachinePool, poolName string, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+
+	// Control plane/master machines use Cluster API which supports data disk security profiles
+	// If using a ManagedDisk the StorageAccountType must be defined
+	if poolName == "master" || poolName == "" { // empty poolName means default machine pool (control plane)
+		for i, dataDisk := range p.DataDisks {
+			if dataDiskContainsSecurityEncryptionType(dataDisk) {
+				allErrs = append(allErrs, field.Invalid(
+					fldPath.Index(i).Child("managedDisk").Child("securityProfile").Child("SecurityEncryptionType"),
+					dataDisk.ManagedDisk.SecurityProfile.SecurityEncryptionType,
+					"security encryption types are not supported on data disks"))
+			}
+
+			if dataDisk.ManagedDisk != nil {
+				if dataDisk.ManagedDisk.StorageAccountType == "" {
+					allErrs = append(allErrs, field.Invalid(
+						fldPath.Index(i).Child("managedDisk").Child("storageAccountType"),
+						dataDisk.ManagedDisk.StorageAccountType,
+						"storageAccount type must not be empty"))
+				}
+			}
+		}
+
+		return allErrs
+	}
+
+	// Worker/compute machines use Machine API which does not support data disk security profiles
+	for i, dataDisk := range p.DataDisks {
+		if dataDiskContainsSecurityEncryptionType(dataDisk) {
+			allErrs = append(allErrs, field.Invalid(
+				fldPath.Index(i).Child("managedDisk").Child("securityProfile").Child("SecurityEncryptionType"),
+				dataDisk.ManagedDisk.SecurityProfile.SecurityEncryptionType,
+				"security encryption types are not supported on data disks"))
+		}
+
+		if dataDisk.ManagedDisk != nil && dataDisk.ManagedDisk.SecurityProfile != nil {
+			allErrs = append(allErrs, field.Invalid(
+				fldPath.Index(i).Child("managedDisk").Child("securityProfile"),
+				dataDisk.ManagedDisk.SecurityProfile,
+				"data disk security profiles are not supported for worker machines (Machine API limitation). Security profiles for data disks are only supported on control plane machines."))
+		}
+	}
+
+	return allErrs
+}
+
+func dataDiskContainsSecurityEncryptionType(dataDisk capz.DataDisk) bool {
+	if dataDisk.ManagedDisk != nil && dataDisk.ManagedDisk.SecurityProfile != nil {
+		if dataDisk.ManagedDisk.SecurityProfile.SecurityEncryptionType != "" {
+			return true
+		}
+	}
+	return false
+}
```

Comment:
```suggestion
func dataDiskContainsSecurityEncryptionType(dataDisk capz.DataDisk) bool {
	return dataDisk.ManagedDisk != nil &&
		dataDisk.ManagedDisk.SecurityProfile != nil &&
		dataDisk.ManagedDisk.SecurityProfile.SecurityEncryptionType != ""
}
```

:bulb: nit: we can simplify it a bit...